pkg/resource/api.go

```diff
@@ -152,6 +153,15 @@ func (a *APIPatchingApplicator) Apply(ctx context.Context, o client.Object, ao .
 	return errors.Wrap(a.client.Patch(ctx, o, &patch{desired}), "cannot patch object")
 }
 
+// WithMergeOptions returns an ApplyOption for merging the value at the given
+// fieldPath of desired object onto the current object with
+// the given merge options.
```

Comment:
I wonder whether we should consider moving to server side apply rather than building our own merge implementation client side?

https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy

Reply:
I will take a look.

Reply:
Hi @negz,
Took a look at server-side apply. My understanding is that it is mainly addressing a somewhat different kind of a problem than the feature we are trying to implement, though we can employ conflict resolution mechanics of server-side apply to achieve merging/replacing fields as you suggested. And this does *not* mean that we cannot change the feature formulation (or that my understanding does not have flaws). 

My understanding is that server-side apply primarily aims a "more stable object lifecycle" by keeping track of appliers (aka fieldManagers) as the managers of a set of fields, and giving a conflict if a field is being modified by another applier than the current fieldManager. We can control how field ownership will be defined, at "compile" time, via merge strategies, e.g., define a merge-strategy of `granular` for a map if we would like to allow different appliers to control different keys of a map, or a merge-strategy of `atomic` if would like the map to be managed by a single applier with all its keys.

With the current implementation here, we are allowing a composition to specify the merge strategy at runtime and it works bidirectionally, i.e., while merging from composite into composed or from composed into composite. One composition can request a different merge strategy than another. 

I'm not sure but server-side apply sounded to me like something of a broader context than adding this feature. For example, I think we may want to reconsider `resource.APIPatchingApplicator` if we decide to employ apply patches and the reconcilers themselves (though we may also choose to use server-side apply only in specific contexts).